auth/auth/auth.py

```diff
@@ -53,13 +76,153 @@ async def get_index(request):  # pylint: disable=unused-argument
     return aiohttp.web.HTTPFound(deploy_config.external_url('auth', '/login'))
 
 
+@routes.get('/signup')
+@web_maybe_authenticated_user
+async def signup(request, userdata):
+    db = request.app['db']
+
+    next = request.query.get('next', deploy_config.external_url('notebook', ''))
+
+    session = await aiohttp_session.get_session(request)
+
+    if 'pending' in session:
+        email = session['email']
+        user = await user_from_email(db, email)
+
+        if user is None:
+            cleanup_session(session)
+            return aiohttp.web.HTTPFound(next)
+
+        if user['state'] == 'deleting' or user['state'] == 'deleted':
+            cleanup_session(session)
+            set_message(session, f'Account for {user["username"]} is deleted.', 'error')
+            return aiohttp.web.HTTPFound(next)
+
+        if user['state'] == 'active':
+            session_id = await create_session(db, user['id'])
+            cleanup_session(session)
+            session['session_id'] = session_id
+            return aiohttp.web.HTTPFound(next)
+
+        assert user['state'] == 'creating'
+
+        page_context = {
+            'username': user['username'],
+            'state': user['state'],
+            'email': user['email']
+        }
+
+        return await render_template('auth', request, userdata, 'signup.html', page_context)
+
+    flow = get_flow(deploy_config.external_url('auth', '/oauth2callback'))
+
+    authorization_url, state = flow.authorization_url(
+        access_type='offline',
+        include_granted_scopes='true')
+
+    session = await aiohttp_session.new_session(request)
+    session['state'] = state
+    session['next'] = next
+    session['caller'] = 'signup'
+
+    return aiohttp.web.HTTPFound(authorization_url)
+
+
+async def _wait_websocket(request, email):
+    app = request.app
+    db = app['db']
+
+    ws = web.WebSocketResponse()
+    await ws.prepare(request)
+
+    user = await user_from_email(db, email)
+    if not user:
+        return web.HTTPNotFound()
```

Comment:
You should do this before you create a web socket. I'm not even sure what happens if you try to send an HTTP response over a connection that has been converted to a web socket.